File: apps/ios/Twende/Twende/Repositories/AuthRepository.swift
import Foundation

final class AuthRepository {
    static let shared = AuthRepository()
    private let api = APIService.shared
    private let tokenStore = TokenStore.shared

    private init() {}

    func login(phone: String, password: String) async throws -> User {
        let body = LoginRequest(phone: phone, password: password)
        let response: APIResponse<AuthData> = try await api.post("auth/login", body: body, authenticated: false)
        guard let data = response.data else {
            throw APIError.serverError(400, response.message ?? "Login failed")
        }
        tokenStore.save(accessToken: data.tokens.accessToken, refreshToken: data.tokens.refreshToken, user: data.user)
        return data.user
    }

    func register(name: String, phone: String, password: String) async throws -> User {
        let body = RegisterRequest(phone: phone, password: password, name: name)
        let response: APIResponse<AuthData> = try await api.post("auth/register", body: body, authenticated: false)
        guard let data = response.data else {
            throw APIError.serverError(400, response.message ?? "Registration failed")
        }
        tokenStore.save(accessToken: data.tokens.accessToken, refreshToken: data.tokens.refreshToken, user: data.user)
        return data.user
    }

    func verifyOtp(phone: String, code: String) async throws {
        let body = OtpVerifyRequest(phone: phone, code: code)
        let _: APIResponse<EmptyResponse> = try await api.post("auth/verify-otp", body: body, authenticated: false)
    }

    func logout() {
        tokenStore.clear()
    }
}

File: apps/android/app/src/main/java/com/twende/app/data/repository/AuthRepository.kt
package com.twende.app.data.repository

import com.twende.app.data.api.TwendeApi
import com.twende.app.data.local.TokenManager
import com.twende.app.data.model.AuthData
import com.twende.app.data.model.LoginRequest
import com.twende.app.data.model.OtpVerifyRequest
import com.twende.app.data.model.RegisterRequest
import com.twende.app.data.model.User
import kotlinx.coroutines.flow.Flow
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AuthRepository @Inject constructor(
    private val api: TwendeApi,
    private val tokenManager: TokenManager,
) {
    suspend fun login(phone: String, password: String): Result<AuthData> = runCatching {
        val response = api.login(LoginRequest(phone, password))
        if (response.isSuccessful) {
            val body = response.body()!!
            if (body.success && body.data != null) {
                persistSession(body.data)
                body.data
            } else {
                throw Exception(body.error?.message ?: body.message ?: "Login failed")
            }
        } else {
            throw Exception("Login failed: ${response.code()}")
        }
    }

    suspend fun register(phone: String, password: String, name: String): Result<AuthData> = runCatching {
        val response = api.register(RegisterRequest(phone, password, name))
        if (response.isSuccessful) {
            val body = response.body()!!
            if (body.success && body.data != null) {
                persistSession(body.data)
                body.data
            } else {
                throw Exception(body.error?.message ?: body.message ?: "Registration failed")
            }
        } else {
            throw Exception("Registration failed: ${response.code()}")
        }
    }

    suspend fun verifyOtp(phone: String, code: String): Result<AuthData> = runCatching {
        val response = api.verifyOtp(OtpVerifyRequest(phone, code))
        if (response.isSuccessful) {
            val body = response.body()!!
            if (body.success && body.data != null) {
                persistSession(body.data)
                body.data
            } else {
                throw Exception(body.error?.message ?: body.message ?: "OTP verification failed")
            }
        } else {
            throw Exception("OTP verification failed: ${response.code()}")
        }
    }

    suspend fun logout(): Result<Unit> = runCatching {
        try {
            api.logout()
        } catch (_: Exception) {
            // Best-effort server logout; always clear local state
        }
        tokenManager.clearAll()
    }

    fun isLoggedIn(): Flow<Boolean> = tokenManager.isLoggedInFlow

    fun getCurrentUser(): Flow<User?> = tokenManager.userFlow

    private suspend fun persistSession(authData: AuthData) {
        tokenManager.saveTokens(
            accessToken = authData.tokens.accessToken,
            refreshToken = authData.tokens.refreshToken,
        )
        tokenManager.saveUser(authData.user)
    }
}
